### PR TITLE
feat(surveys): preact surveys components

### PR DESCRIPTION
**Pull request by liyiy**
## Changes

Converts surveys into components

How to manually test:

1. Follow the instructions for linking this version of posthog-js to your main posthog branch https://github.com/PostHog/posthog-js?tab=readme-ov-file#developing-together-with-another-project

2. Test any and all flows including:

- [ ] All survey types
- [ ] Single and multiple choice open ended select
- [ ] Multiple question surveys
- [ ] Customization with colors, branding, html in description, etc.
- [ ] Survey results look correct
- [ ] Feedback tab and custom selector widget

## Checklist
- [ ] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers

### CODE REVIEW

**Comment by liyiy** [comment]
I added a bunch of e2e tests and also manually went through the flows, but would very much appreciate a second or third look from others! 🙏

**Comment by neilkakkar** [comment]
Great job adding all the e2e tests! 🚀 

Suggestion: Split e2e tests into a separate PR, merge it first, so we can confirm they work well with both the old and new code

File: src/extensions/surveys.tsx [hunk 1]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+    onSubmit,
+    link,
+}: {
+    text: string
+    submitDisabled: boolean
+    appearance: SurveyAppearance
+    onSubmit: () => void
+    link?: string | null
+}) {
+    const backgroundColorRef = Preact.createRef()
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:164]: Why is this `createRef`? I would have assumed `useRef()` would be the functional way of working? Otherwise that create call is happening on every render which means it doesn't feed into the hook system. 

Same applies for all uses of it I can see 🤔 

Also the ref is to the button not the color

File: src/extensions/surveys.tsx [hunk 2]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
         callSurveys(posthog, false)
     }, 3000)
 }
+
+const defaultBackgroundColor = '#eeeded'
+
+export function QuestionHeader({
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:124]: Generally speaking I would already want this to be a bit cleaner as a codebase. No reason I can see why not to put this in the `surveys` subfolders something like:

```
/surveys/components/QuestionHeader.tsx
/surveys/constants.ts // for things like colors
```

I think it's worth doing already as it will become messy quickly and harder to review in the future.

File: src/extensions/surveys.tsx [hunk 3]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+    description,
+    backgroundColor,
+}: {
+    question: string
+    description?: string | null
+    backgroundColor?: string
+}) {
+    return (
+        <div style={{ backgroundColor: backgroundColor || defaultBackgroundColor }}>
+            <div className="survey-question">{question}</div>
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:135]: I understand this is out of the scope of this PR but for long term code quality it would be nice to have a solid base of BEM style classes rather than random simple names.

Not expecting this to happen now but follow up ideas...

File: src/extensions/surveys.tsx (hunk 1 above)
Comment (benjackwhite) [reply src/extensions/surveys.tsx:164]:
```suggestion
    const buttonRef = useRef<HTMLButtonElement>(null)
```

File: src/extensions/surveys.tsx [hunk 4]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+    onSubmit: () => void
+    link?: string | null
+}) {
+    const backgroundColorRef = Preact.createRef()
+    const [textColor, setTextColor] = useState('white')
+
+    useEffect(() => {
+        const color = getTextColor(backgroundColorRef.current)
+        setTextColor(color)
+    }, [appearance])
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:170]: This whole block was pretty confusing 😅 

I sort of get what it does I think - using the element to find the color and then setting an appropriate text color.

Feels like this would be a strong use case for `useContext` passing the appearance object through so that you can determine it without having to query the DOM inefficiently? 

For now I guess it's fine but could be a good thing to follow up on as this could be an easy source of rendering bugs in the future.

File: src/extensions/surveys.tsx [hunk 5]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+        </div>
+    )
+}
+
+export function PostHogLogo({ backgroundColor }: { backgroundColor?: string }) {
+    const backgroundColorRef = Preact.createRef()
+    const [textColor, setTextColor] = useState('white')
+    useEffect(() => {
+        setTextColor(getTextColor(backgroundColorRef.current))
+    }, [])
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:203]: This block is almost identical across multiple components.

Could be turned into a custom hook that encapsulates it:

```tsx
export function useContrastingTextColor(options: { appearance: SurveyAppearance; defaultTextColor?: string }): {
    ref: React.RefObject<HTMLElement>
    textColor: string
} {
    const ref = useRef<HTMLElement>(null)
    const [textColor, setTextColor] = useState(options.defaultTextColor ?? 'white')

    useEffect(() => {
        if (ref.current) {
            const color = getTextColor(ref.current)
            setTextColor(color)
        }
    }, [options.appearance])

    return {
        ref,
        textColor,
    }
}


function MyComponent() {
      const { textColor, ref } = useContrastingTextColor({ appearance })
}


```

File: src/extensions/surveys.tsx [hunk 6]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+
+    return (
+        <a
+            href="https://posthog.com"
+            target="_blank"
+            rel="noopener"
+            ref={backgroundColorRef}
+            style={{ backgroundColor: backgroundColor, color: textColor }}
+            className="footer-branding auto-text-color"
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:212]: What does `auto-text-color` do? Couldn't find any reference to it.

File: src/extensions/surveys.tsx [hunk 7]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+                    confirmationHeader={survey.appearance?.thankYouMessageHeader || 'Thank you!'}
+                    confirmationDescription={survey.appearance?.thankYouMessageDescription || ''}
+                    appearance={survey.appearance || {}}
+                    styleOverrides={{ ...style, ...confirmationBoxLeftStyle }}
+                    onClose={() => setShowConfirmation(false)}
+                />
+            )}
+        </>
+    )
+}
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:262]: There is a potential for bugs here. This is more of a simple state machine so you don't want multiple sources of truth but rather one option that decides what is shown. Something like: 

```suggestion
export function Surveys({ posthog, survey, style }: { posthog: PostHog; survey: Survey; style?: React.CSSProperties }) {
    const [displayState, setDisplayState] = useState<'survey' | 'confirmation' | 'closed'>('survey')

    useEffect(() => {
        window.dispatchEvent(new Event('PHSurveyShown'))
        posthog.capture('survey shown', {
            $survey_name: survey.name,
            $survey_id: survey.id,
            sessionRecordingUrl: posthog.get_session_replay_url?.(),
        })
        localStorage.setItem(`lastSeenSurveyDate`, new Date().toISOString())

        window.addEventListener('PHSurveyClosed', () => {
            setDisplayState('closed')
        })

        window.addEventListener('PHSurveySent', () => {
            setDisplayState('confirmation')
            if (survey.appearance?.autoDisappear) {
                setTimeout(() => {
                    setDisplayState('closed')
                }, 5000)
            }
        })
    }, [])
    const confirmationBoxLeftStyle = style?.left && _isNumber(style?.left) ? { left: style.left - 40 } : {}

    return (
        <>
            {displayState === 'survey' && <Questions survey={survey} posthog={posthog} styleOverrides={style} />}
            {displayState === 'confirmation' && survey.appearance?.displayThankYouMessage && (
                <ConfirmationMessage
                    confirmationHeader={survey.appearance?.thankYouMessageHeader || 'Thank you!'}
                    confirmationDescription={survey.appearance?.thankYouMessageDescription || ''}
                    appearance={survey.appearance || {}}
                    styleOverrides={{ ...style, ...confirmationBoxLeftStyle }}
                    onClose={() => setDisplayState('closed')}
                />
            )}
        </>
    )
}
```

File: src/extensions/surveys.tsx [hunk 8]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+        $survey_name: survey.name,
+        $survey_id: survey.id,
+        sessionRecordingUrl: posthog.get_session_replay_url?.(),
+        $set: {
+            [`$survey_dismissed/${survey.id}`]: true,
+        },
+    })
+    localStorage.setItem(`seenSurvey_${survey.id}`, 'true')
+    window.dispatchEvent(new Event('PHSurveyClosed'))
+}
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:340]: It would be great if at some point we could move this into propert state management. We don't have to go full Kea.js but some nicer way of encapsulating this logic would be good (also so that we could add unit tests)

File: src/extensions/surveys.tsx [hunk 9]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+    question: RatingSurveyQuestion
+    questionIndex: number
+    appearance: SurveyAppearance
+    onSubmit: (rating: number | null) => void
+    closeSurveyPopup: () => void
+}) {
+    const threeScaleEmojis = [dissatisfiedEmoji, neutralEmoji, dissatisfiedEmoji]
+    const fiveScaleEmojis = [veryDissatisfiedEmoji, dissatisfiedEmoji, neutralEmoji, satisfiedEmoji, verySatisfiedEmoji]
+    const fiveScaleNumbers = [1, 2, 3, 4, 5]
+    const tenScaleNumbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:420]: These are all constants that should be moved to a constants.ts file or something. It's not the biggest deal in the world that these are declared in each function call but it's far from best practice.

File: src/extensions/surveys.tsx [hunk 10]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+                        choiceClass += ' choice-option-open'
+                    }
+                    return (
+                        <div className={choiceClass}>
+                            <input
+                                type={inputType}
+                                id={`surveyQuestion${questionIndex}Choice${idx}`}
+                                name={`question${questionIndex}`}
+                                value={val}
+                                // disabled={!choice}
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:620]: ?

File: src/extensions/surveys.tsx [hunk 11]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+                                            question.type === SurveyQuestionType.MultipleChoice &&
+                                            _isArray(selectedChoices)
+                                        ) {
+                                            if (selectedChoices.includes(val)) {
+                                                setSelectedChoices(selectedChoices.filter((choice) => choice !== val))
+                                            } else {
+                                                setSelectedChoices([...selectedChoices, val])
+                                            }
+                                        }
+                                    }
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:637]: My eyes 🙈 !!

Nested ifs are another source of bugs as well as being hard to parse. This is basically a switch statement so its way nicer to use returns to break the logic rather than having deeper and deeper nesting.

```suggestion
                                    if (question.hasOpenChoice && idx === question.choices.length - 1) {
                                        return setOpenChoiceSelected(!openChoiceSelected)
                                    }
                                    if (question.type === SurveyQuestionType.SingleChoice) {
                                        return setSelectedChoices(val)
                                    }

                                    if (
                                        question.type === SurveyQuestionType.MultipleChoice &&
                                        _isArray(selectedChoices)
                                    ) {
                                        if (selectedChoices.includes(val)) {
                                            return setSelectedChoices(
                                                selectedChoices.filter((choice) => choice !== val)
                                            )
                                        }
                                        return setSelectedChoices([...selectedChoices, val])
                                    }
```

Comment (benjackwhite) [reply src/extensions/surveys.tsx:637]: Furthermore - as soon as the logic is more than a one-liner or so you should pull it out to its own function above the JSX. Otherwise the markup becomes even more unreadable.

File: src/extensions/surveys.tsx [hunk 12]
```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
...
+                                            onInput={(e) => {
+                                                const userValue = e.currentTarget.value
+                                                if (question.type === SurveyQuestionType.SingleChoice) {
+                                                    setSelectedChoices(userValue)
+                                                } else if (
+                                                    question.type === SurveyQuestionType.MultipleChoice &&
+                                                    _isArray(selectedChoices)
+                                                ) {
+                                                    setOpenEndedInput(userValue)
+                                                }
```
Comment (benjackwhite) [inline src/extensions/surveys.tsx:657]: Same comment about nested ifs here

**Review by benjackwhite — CHANGES_REQUESTED** [review]
Bunch of comments, mostly about code quality rather than the functionality itself. I think its worth the extra effort to make this a little tidier now (otherwise it will likely not happen 🙈 )

File: src/extensions/surveys.tsx (hunk 9 above)
Comment (liyiy) [reply src/extensions/surveys.tsx:420]: I wasn't able to pull these into a constants file because of the Preact separate components issue as commented above, but I did take them out of the function call.

File: src/extensions/surveys.tsx (hunk 8 above)
Comment (liyiy) [reply src/extensions/surveys.tsx:340]: added as a TODO

File: src/extensions/surveys.tsx (hunk 4 above)
Comment (liyiy) [reply src/extensions/surveys.tsx:170]: added as a TODO

**Comment by liyiy** [comment]
Thanks @benjackwhite for the thorough review. I've either added in the fixes or left the less urgent ones as TODOs. Let me know if I've missed anything !

File: src/extensions/surveys.tsx (hunk 2 above)
Comment (benjackwhite) [reply src/extensions/surveys.tsx:124]: Not following the reasoning tbh? You need to import preact, therefore you don't put them in different files?

**Review by benjackwhite — APPROVED** [review]
This is much nicer! Easy to review, easy to see how one would make changes later. This was definitely worth the extra investment from my perspective.

Haven't tested the functionality itself - I would watch this like a hawk after it gets merged and ensure that there are no surprises...

**Review by neilkakkar — APPROVED** [review]
🚀
